src/Microsoft.Azure.SignalR.Common/ServiceConnectionContainer.cs

```diff
@@ -0,0 +1,110 @@
+    {
+        private const int RetryCount = 10;
+        private readonly List<IServiceConnection> _serviceConnections;
+        private readonly int _count;
+
+        public ServiceConnectionContainer(List<IServiceConnection> serviceConnections)
+        {
+            _serviceConnections = serviceConnections;
+            _count = _serviceConnections.Count;
+        }
+
+        public ServiceConnectionContainer(Func<IServiceConnection> generator, int count)
+        {
+            if (generator == null)
+            {
+                throw new ArgumentNullException(nameof(generator));
+            }
+
+            if (count <= 0)
+            {
+                throw new ArgumentException($"{nameof(count)} must be greater than 0.");
+            }
+
+            _serviceConnections = new List<IServiceConnection>(count);
+            for (int i = 0; i < count; i++)
+            {
+                _serviceConnections.Add(generator());
+            }
+
+            _count = _serviceConnections.Count;
+        }
+
+        public ServiceConnectionStatus Status => throw new NotSupportedException();
+
+        public Task StartAsync()
+        {
+            return Task.WhenAll(_serviceConnections.Select(c => c.StartAsync()));
+        }
+
+        public Task StopAsync()
+        {
+            return Task.WhenAll(_serviceConnections.Select(c => c.StopAsync()));
+        }
+
+        public Task WriteAsync(ServiceMessage serviceMessage)
+        {
+            return WriteToRandomAvailableConnection(serviceMessage);
+        }
+
+        public Task WriteAsync(string partitionKey, ServiceMessage serviceMessage)
+        {
+            // If we hit this check, it is a code bug.
+            if (string.IsNullOrEmpty(partitionKey))
+            {
+                throw new ArgumentNullException(nameof(partitionKey));
+            }
+
+            return WriteToPartitionedConnection(partitionKey, serviceMessage);
+        }
+
+        private Task WriteToPartitionedConnection(string partitionKey, ServiceMessage serviceMessage)
+        {
+            return WriteWithRetry(serviceMessage, i => Math.Abs((partitionKey.GetHashCode() + i) % _count), _count);
+        }
+
+        private Task WriteToRandomAvailableConnection(ServiceMessage sm)
+        {
+            return WriteWithRetry(sm, i => StaticRandom.Next(_count), RetryCount);
+        }
+
+        private async Task WriteWithRetry(ServiceMessage sm, Func<int, int> nextSelector, int maxRetry)
+        {
+            var retry = 0;
+            while (retry < maxRetry)
+            {
+                var index = nextSelector(retry);
+                var connection = _serviceConnections[index];
+                if (connection != null && connection.Status == ServiceConnectionStatus.Connected)
+                {
+                    try
+                    {
+                        // still possible the connection is not valid
```

Comment:
>// still possible the connection is not valid [](start = 24, length = 45)

Just question, when would Status set not accurate?

Reply:
because the check is not atomic